cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -0,0 +1,189 @@
+        try {
+          for (RecordWriter<K, V> recordWriter : this.recordWriters.values()) {
+            recordWriter.close(context);
+          }
+          this.recordWriters.clear();
+
+          mrTaskContext.flushOperations();
+        } catch (Exception e) {
+          throw new IOException(e);
+        } finally {
+          mrTaskContext.close();
+        }
+      }
+    };
+  }
+
+  private static TaskAttemptContext getTaskAttemptContext(TaskAttemptContext context,
+                                                          String newOutputName) throws IOException {
+    Job job = new Job(context.getConfiguration());
+    FileOutputFormat.setOutputName(job, newOutputName);
+
+    Path jobOutputPath = createJobSpecificPath(FileOutputFormat.getOutputPath(job), context);
+    FileOutputFormat.setOutputPath(job, jobOutputPath);
+
+    return new TaskAttemptContextImpl(job.getConfiguration(), context.getTaskAttemptID());
+  }
+
+  /**
+   * @return A RecordWriter object over the given file
+   * @throws IOException
+   */
+  protected RecordWriter<K, V> getBaseRecordWriter(TaskAttemptContext job) throws IOException, InterruptedException {
+    if (fileOutputFormat == null) {
+      // TODO: use the PFS's OutputFormat (encode in conf)
+      fileOutputFormat = new TextOutputFormat<>();
+    }
+    return fileOutputFormat.getRecordWriter(job);
+  }
+
+  // suffixes a Path with a job-specific string
+  private static Path createJobSpecificPath(Path path, JobContext jobContext) {
+    String outputPathSuffix = "_temporary_" + jobContext.getJobID().getId();
+    return new Path(path, outputPathSuffix);
+  }
+
+  @Override
+  public synchronized FileOutputCommitter getOutputCommitter(TaskAttemptContext context) throws IOException {
+    if (committer == null) {
+      final Path jobSpecificOutputPath = createJobSpecificPath(getOutputPath(context), context);
+      committer = new DynamicPartitioningOutputCommitter(jobSpecificOutputPath, context);
+    }
+    return committer;
+  }
+
+  @Override
+  public void checkOutputSpecs(JobContext job) throws IOException {
+    // Ensure that the output directory is set and not already there
+    Path outDir = getOutputPath(job);
+    if (outDir == null) {
+      throw new InvalidJobConfException("Output directory not set.");
+    }
+
+    // get delegation token for outDir's file system
+    TokenCache.obtainTokensForNamenodes(job.getCredentials(),
+                                        new Path[]{outDir}, job.getConfiguration());
+
+    // we permit multiple jobs writing to the same output directory. We handle this by each one writing to distinct
+    // paths within that directory. See createJobSpecificPath method and usages of it.
+
+    // additionally check that output dataset and dynamic partitioner class name has been set in conf
+    if (job.getConfiguration().get(HCONF_ATTR_OUTPUT_DATASET) == null) {
+      throw new InvalidJobConfException("Output dataset not set.");
```

Comment:
A better error message would be "The job configuration does not contain required property " + HCONF_ATTR_OUTPUT_DATASET